presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java

```
@@ -0,0 +1,227 @@
+                    if (data.get(i).size() >= 2) {
+                        String tableId = String.valueOf(data.get(i).get(0));
+                        String sheetId = String.valueOf(data.get(i).get(1));
+                        tableSheetMap.put(tableId, Optional.of(sheetId));
+                    }
+                }
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_METASTORE_ERROR, e);
+        }
+        return tableSheetMap;
+    }
+
+    private Optional<List<List<Object>>> getDataForSheetExpression(String sheetExpression)
+    {
+        try {
+            return Optional.of(readAllValuesFromSheetExpression(sheetExpression));
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_UNKNOWN_TABLE_ERROR, e);
+        }
+    }
+
+    private Credential getCredentials() throws IOException
+    {
+        GoogleCredential credential;
+        // Load client secrets.
+        try (InputStream in = new FileInputStream(credentialsFilePath)) {
+            credential = GoogleCredential.fromStream(in).createScoped(SCOPES);
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
+        }
+        return credential;
+    }
+
+    public Set<String> listTables()
+    {
+        Set<String> tables = new HashSet<>();
+        try {
+            List<List<Object>> tableMetadata = sheetDataCache.get(metadataSheetId).get();
+            if (tableMetadata.size() > 1) { // we need both table name and sheet id
+                for (int i = 1; i < tableMetadata.size(); i++) {
+                    if (tableMetadata.get(i).size() > 0) {
+                        tables.add(String.valueOf(tableMetadata.get(i).get(0)));
+                    }
+                }
+            }
+        }
+        catch (Exception e) {
+            log.error(e, "Error retrieving table metadata mapping");
```

please do not swallow exceptions